Mage.Sets/src/mage/cards/d/DihadaBinderOfWills.java

```diff
@@ -0,0 +1,211 @@
+        ability.addTarget(new TargetCreaturePermanent(0, 1, legendarycreaturefilter, false));
+        this.addAbility(ability);
+
+
+        // -3: Reveal the top four cards of your library. Put any number of legendary cards from among them into your hand and the rest into your graveyard. Create a Treasure token for each card put into your graveyard this way.
+        effect = new DihadaFilterEffect();
+        ability = new LoyaltyAbility(effect, -3);
+        this.addAbility(ability);
+
+        // -11: Gain control of all nonland permanents until end of turn. Untap them. They gain haste until end of turn.
+        effect = new DihadaControlEffect();
+        ability = new LoyaltyAbility(effect, -11);
+        this.addAbility(ability);
+        
+        // Dihada, Binder of Wills can be your commander.
+        this.addAbility(CanBeYourCommanderAbility.getInstance());
+    }
+
+    private DihadaBinderOfWills(final DihadaBinderOfWills card) {
+        super(card);
+    }
+
+    @Override
+    public DihadaBinderOfWills copy() {
+        return new DihadaBinderOfWills(this);
+    }
+}
+
+class DihadaFilterEffect extends OneShotEffect {
+
+    private static final FilterCard legendaryfilter = new FilterCard("legendary card");
+
+    static {
+        legendaryfilter.add(SuperType.LEGENDARY.getPredicate());
+    }
+
+    public DihadaFilterEffect() {
+        super(Outcome.PutCardInHand);
+        staticText = "Reveal the top four cards of your library. Put any "
+                + "number of legendary cards from among them "
+                + "into your hand and the rest into your graveyard. "
+                + "Create a treasure token for each card put into "
+                + "your graveyard this way..";
+    }
+
+    public DihadaFilterEffect(final DihadaFilterEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        MageObject sourceObject = game.getObject(source);
+        if (controller == null || sourceObject == null) {
+            return false;
+        }
+        Cards cards = new CardsImpl(controller.getLibrary().getTopCards(game, 4));
+        if (!cards.isEmpty()) {
+            controller.revealCards(sourceObject.getIdName(), cards, game);
+
+            TargetCard target1 = new TargetCard(0, Integer.MAX_VALUE, Zone.LIBRARY, legendaryfilter);
+            target1.setNotTarget(true);
+            controller.choose(Outcome.PutCardInHand, cards, target1, game);
+            Cards toHand = new CardsImpl(target1.getTargets());
+            cards.removeAll(toHand);
+            controller.moveCards(toHand.getCards(game), Zone.HAND, source, game);
+            controller.moveCards(cards, Zone.GRAVEYARD, source, game);
+
+            //Make Treasure
+            new TreasureToken().putOntoBattlefield(cards.size(), game, source, source.getControllerId());
```

Comment:
You may need to check against `cards.size() == 0`. I don't know it works correctly when 0 is passed in.

Reply:
Possibly, but better safe than sorry, added.

Reply:
I would prefer that this be implemented by subclassing `RevealLibraryPickControllerEffect` rather than implementing the effect from scratch. Look at Discerning Taste for an example (that class subclasses `LookLibraryAndPickControllerEffect`, but `RevealLibraryPickControllerEffect` is a very thin extension of that class and can be extended in exactly the same way).

The only reason GenesisWave uses a custom effect is because the filter dynamically depends on X.

Reply:
This also has to create treasures based on how many cards are left, and I have no clue how the other function works.
Genesis wave had both the voluntary choice, and a list of the remaining cards (to count up for the treasures)

Reply:
Did you look at Discerning Taste? What you do is override `actionWithPickedCards(Game game, Ability source, Player player, Cards pickedCards, Cards otherCards)` In your override, you first call `super.actionWithPickedCards` which does the card moving. Then do whatever you want with `pickedCards` and/or `otherCards` (in this case, create Treasures equal to `otherCards.size()`).

Reply:
Lacks the legendary filter, and the variable number (from what i can see)

Reply:
All right, how would you add this ability to the card if it didn't have the last sentence? You would not need a custom `Effect` class at all, you would just use a common class:

`RevealLibraryPickControllerEffect(4, Integer.MAX_VALUE, filter, PutCards.HAND, PutCards.GRAVEYARD, false)`

4 is the number of cards to look at.
`Integer.MAX_VALUE` is the maximum number of cards that can be picked ("any number")
`PutCards.HAND` is what to do with the cards the player picked.
`PutCards.GRAVEYARD` is what to do with the cards the player didn't pick.
`false` because picking the cards is not optional (it doesn't say "**you may** put")

So you construct your custom subclass using the above arguments to `super`:
```
class DihadaFilterEffect extends RevealLibraryPickControllerEffect {

    DihadaFilterEffect() {
        super(4, Integer.MAX_VALUE, filter, PutCards.HAND, PutCards.GRAVEYARD, false);
        staticText = blah blah
    }
```

Reply:
Funny thing, "Put any number" includes 0, so it is sort of a may.

Reply:
Yes, but if you don't include that argument, the common effect code will redundantly ask the player if they want to pick any cards, and then give them the target interface with the option of clicking "done" with no cards picked. The constructor is fairly pedantic because there are a lot of versions of this ability with subtly different templating.